Sources/Crypto/Digests/BoringSSL/Digest_boring.swift

```diff
@@ -67,81 +141,44 @@ struct OpenSSLDigestImpl<H: BoringSSLBackedHashFunction> {
     }
 
     internal func finalize() -> H.Digest {
-        // To have a non-destructive finalize operation we must allocate.
-        let copyContext = DigestContext(copying: self.context)
-        let digestBytes = copyContext.finalize()
-        return digestBytes.withUnsafeBytes {
+        return self.context.finalize {
             // We force unwrap here because if the digest size is wrong it's an internal error.
             H.Digest(bufferPointer: $0)!
         }
     }
 }
 
-class DigestContext {
-    private var contextPointer: UnsafeMutablePointer<EVP_MD_CTX>
+final class DigestContext<H: BoringSSLBackedHashFunction> {
+    @exclusivity(unchecked) private var context: H.Context
 
-    init(digest: DigestType) {
-        // We force unwrap because we cannot recover from allocation failure.
-        self.contextPointer = CCryptoBoringSSL_EVP_MD_CTX_new()!
-        guard CCryptoBoringSSL_EVP_DigestInit(self.contextPointer, digest.dispatchTable) != 0 else {
-            // We can't do much but crash here.
-            fatalError("Unable to initialize digest state: \(CCryptoBoringSSL_ERR_get_error())")
-        }
+    init() {
+        self.context = H.initialize()
     }
 
     init(copying original: DigestContext) {
-        // We force unwrap because we cannot recover from allocation failure.
-        self.contextPointer = CCryptoBoringSSL_EVP_MD_CTX_new()!
-        guard CCryptoBoringSSL_EVP_MD_CTX_copy(self.contextPointer, original.contextPointer) != 0 else {
-            // We can't do much but crash here.
-            fatalError("Unable to copy digest state: \(CCryptoBoringSSL_ERR_get_error())")
-        }
+        self.context = original.context
     }
 
     func update(data: UnsafeRawBufferPointer) {
-        guard let baseAddress = data.baseAddress else {
-            return
-        }
-
-        CCryptoBoringSSL_EVP_DigestUpdate(self.contextPointer, baseAddress, data.count)
+        H.update(&self.context, data: data)
     }
 
-    // This finalize function is _destructive_: do not call it if you want to reuse the object!
-    func finalize() -> [UInt8] {
-        let digestSize = CCryptoBoringSSL_EVP_MD_size(self.contextPointer.pointee.digest)
-        var digestBytes = Array(repeating: UInt8(0), count: digestSize)
-        var count = UInt32(digestSize)
-
-        digestBytes.withUnsafeMutableBufferPointer { digestPointer in
-            assert(digestPointer.count == count)
-            CCryptoBoringSSL_EVP_DigestFinal(self.contextPointer, digestPointer.baseAddress, &count)
+    func finalize<R>(_ body: (UnsafeRawBufferPointer) -> R) -> R {
```

Comment:
I think we can avoid the extra generic here and just return `H.Digest`, as we already have the generic type parameter.

Reply:
It's done. 🙂